contracts/DivviRegistry.sol

```diff
@@ -244,19 +299,127 @@ contract DivviRegistry is
+   * has already verified the content of `offchainMessage` to ensure it represents a valid referral intent (e.g., contains a specific tag).
+   * A future upgrade will incorporate this content validation fully on-chain.
+   * @param referrals Array of referral data to register
+   */
+  function batchRegisterReferral(
+    ReferralDataV2[] calldata referrals
+  ) external onlyRole(REFERRAL_REGISTRAR_ROLE) {
+    for (uint256 i = 0; i < referrals.length; i++) {
+      ReferralDataV2 calldata referral = referrals[i];
+
+      // Determine if this is an off-chain signature-based referral
+      bool isSignatureBased = referral.offchainMessage.signature.length > 0;
+
+      // Set event values based on referral type
+      string memory eventChainId = isSignatureBased
+        ? 'offchain'
+        : referral.onchainTx.chainId;
+      bytes32 eventTxHash = isSignatureBased
+        ? bytes32(0)
+        : referral.onchainTx.txHash;
+
+      // Process the referral (including signature verification if applicable)
+      ReferralStatus status = _registerReferral(
+        referral.user,
+        referral.rewardsProvider,
+        referral.rewardsConsumer,
+        referral.offchainMessage
+      );
+
+      // Emit appropriate event based on status
+      if (status == ReferralStatus.SUCCESS) {
+        emit ReferralRegistered(
+          referral.user,
+          referral.rewardsProvider,
+          referral.rewardsConsumer,
+          eventChainId,
+          eventTxHash
+        );
+      } else {
+        emit ReferralSkipped(
+          referral.user,
+          referral.rewardsProvider,
+          referral.rewardsConsumer,
+          eventChainId,
+          eventTxHash,
+          status
+        );
+      }
+    }
+  }
+
+  /**
+   * @notice Verify a signature for referral consent
+   * @param user The address that should have signed (EOA or EIP-1271 contract)
+   * @param offchainMessage The off-chain message data containing message, signature, and type
+   * @return valid Whether the signature is valid
+   */
+  function _verifyOffchainReferralSignature(
+    address user,
+    OffchainMessageData memory offchainMessage
+  ) internal view returns (bool valid) {
+    bytes32 messageHash;
+
+    if (offchainMessage.messageType == OffchainMessageType.ETH_SIGNED_MESSAGE) {
+      // Create message hash (Ethereum signed message format)
+      messageHash = MessageHashUtils.toEthSignedMessageHash(
+        offchainMessage.message
+      );
+    } else {
+      // Invalid message type for signature verification
+      return false;
+    }
+
+    // Check if user is a contract (potential smart wallet)
+    if (user.code.length > 0) {
+      // Try EIP-1271 verification for smart contracts
+      try
+        IERC1271(user).isValidSignature(messageHash, offchainMessage.signature)
+      returns (bytes4 magicValue) {
```

Comment:
Just for my own knowledge, what's the deal with this `EIP1271_MAGIC_VALUE` const? I see [this here](https://docs.uniswap.org/contracts/v3/reference/periphery/interfaces/external/IERC1271), but I guess I'm a bit confused as to _why_ this function has this sort of interface.